src/Microsoft.OData.Core/ODataAsynchronousReader.cs

```diff
@@ -302,5 +304,160 @@ private int ReadByte()
+        /// </returns>
+        private async Task<Dictionary<string, string>> ReadHeadersAsync()
+        {
+            var headers = new Dictionary<string, string>(StringComparer.Ordinal);
+
+            // Read all the headers
+            string headerLine = await this.ReadLineAsync()
+                .ConfigureAwait(false);
+
+            while (!string.IsNullOrEmpty(headerLine))
+            {
+                string headerName, headerValue;
+                ValidateHeaderLine(headerLine, out headerName, out headerValue);
+
+                if (headers.ContainsKey(headerName))
+                {
+                    throw new ODataException(Strings.ODataAsyncReader_DuplicateHeaderFound(headerName));
+                }
+
+                headers.Add(headerName, headerValue);
+                headerLine = await this.ReadLineAsync()
+                    .ConfigureAwait(false);
+            }
+
+            return headers;
+        }
+
+        /// <summary>
+        /// Asynchronously reads a line from the underlying stream.
+        /// </summary>
+        /// <returns>
+        /// A task that represents the asynchronous read operation.
+        /// The value of the TResult parameter contains the line read from the stream.
+        /// </returns>
+        private async Task<string> ReadLineAsync()
+        {
+            StringBuilder lineBuilder = new StringBuilder();
+
+            int ch = await this.ReadByteAsync()
+                .ConfigureAwait(false);
+
+            while (ch != -1)
+            {
+                if (ch == '\n')
+                {
+                    throw new ODataException(Strings.ODataAsyncReader_InvalidNewLineEncountered('\n'));
+                }
+
+                if (ch == '\r')
+                {
+                    ch = await this.ReadByteAsync()
+                        .ConfigureAwait(false);
+
+                    if (ch != '\n')
+                    {
+                        throw new ODataException(Strings.ODataAsyncReader_InvalidNewLineEncountered('\r'));
+                    }
+
+                    return lineBuilder.ToString();
+                }
+
+                lineBuilder.Append((char)ch);
+                ch = await this.ReadByteAsync()
+                    .ConfigureAwait(false);
+            }
+
+            throw new ODataException(Strings.ODataAsyncReader_UnexpectedEndOfInput);
+        }
+
+        /// <summary>
+        /// Asynchronously reads a byte from the underlying stream.
+        /// </summary>
+        /// <returns>
+        /// A task that represents the asynchronous read operation.
+        /// The value of the TResult parameter contains the byte read from the stream.
+        /// </returns>
+        private async Task<int> ReadByteAsync()
+        {
+            // Stream does not support ReadByteAsync - Improvise
+            if (await this.rawInputContext.Stream.ReadAsync(byteBuffer, 0, 1).ConfigureAwait(false) != 0)

```

Comment:
If we're reading into a byte buffer directly, maybe it would make sense to return the read byte as `byte`. Or this could return a boolean then the caller reads `byteBuffer[0]` if the return was successful. That way we avoid `byte`->`int` or `byte`->`char` comparisons/conversions. But you could verify that improves things before applying the suggestion.